db/db_impl/db_impl_write.cc

```diff
@@ -75,12 +75,13 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
   if (my_batch == nullptr) {
     return Status::Corruption("Batch is nullptr!");
   }
-  if (tracer_) {
+  if (tracer_ && !tracer_->IsWriteOrderPreserved()) {
+    // We don't have to preserve write order so can trace anywhere. It's more
+    // efficient to trace here than to add latency to a phase of the log/apply
+    // pipeline.
     InstrumentedMutexLock lock(&trace_mutex_);
-    if (tracer_) {
```

Comment:
Is it possible that after this thread get the lock, the tracer is freed? That's why we check tracer again here

Reply:
Thanks I will restore the two checks.

Now that I understand what it's supposed to be doing, is checking `tracer_` without `trace_mutex_` held (L78) a race condition?

Reply:
> Thanks I will restore the two checks.

Done.

> Now that I understand what it's supposed to be doing, is checking tracer_ without trace_mutex_ held (L78) a race condition?

I guess this is an existing issue; doesn't need to be fixed by this PR. I added TODOs though.

Reply:
> > Thanks I will restore the two checks.
> 
> Done.
> 
> > Now that I understand what it's supposed to be doing, is checking tracer_ without trace_mutex_ held (L78) a race condition?
> 
> I guess this is an existing issue; doesn't need to be fixed by this PR. I added TODOs though.
Thanks for adding the comments.